clash-prelude/src/Clash/Explicit/Reset.hs

```diff
@@ -180,23 +184,96 @@ resetSynchronizer clk rst = rstOut
 -- See: https://github.com/clash-lang/clash-compiler/pull/2511
 {-# CLASH_OPAQUE resetSynchronizer #-} -- Give reset synchronizer its own HDL file
 
+resetGlitchFilter, unsafeResetGlitchFilter
+  :: forall dom glitchlessPeriod n
+   . ( HasCallStack
+     , KnownDomain dom
+     , glitchlessPeriod ~ (n + 1) )
+  => SNat glitchlessPeriod
+  -- ^ Consider a reset signal to be properly asserted after having seen the
+  -- reset asserted for /glitchlessPeriod/ cycles straight.
+  -> Clock dom
+  -> Reset dom
+  -> Reset dom
+
 -- | Filter glitches from reset signals by only triggering a reset after it has
 -- been asserted for /glitchlessPeriod/ cycles. Similarly, it will stay
 -- asserted until a /glitchlessPeriod/ number of deasserted cycles have been
 -- observed.
 --
--- This circuit can only be used on platforms supporting initial values. Platforms
--- not supporting this should consider using a power-on-reset and a circuit
--- designed with this environment in mind.
+-- This circuit can only be used on platforms supporting initial values, and
+-- will refuse simulation and HDL generation in domains that have no initial
+-- values ('Unknown').
+--
+-- This check can be worked around by using 'unsafeResetGlitchFilter' but this
+-- is not recommended.
+--
+-- On platforms without initial values, you should instead use
+-- 'resetGlitchFilterWithReset' with an additional power-on reset, or
+-- 'holdReset' if filtering is only needed on deassertion.
+--
+-- At power-on, the reset will be asserted. If the filtered reset input remains
+-- unasserted, the output reset will deassert after /glitchlessPeriod/ clock
+-- cycles.
 --
 -- === __Example 1__
 -- >>> let sampleResetN n = sampleN n . unsafeToActiveHigh
 -- >>> let resetFromList = unsafeFromActiveHigh . fromList
 -- >>> let rst = resetFromList [True, True, False, False, True, False, False, True, True, False, True, True]
 -- >>> sampleResetN 12 (resetGlitchFilter d2 systemClockGen rst)
 -- [True,True,True,True,False,False,False,False,False,True,True,True]
+resetGlitchFilter =
+  case initBehavior @dom of
+    SDefined -> unsafeResetGlitchFilter
+    SUnknown -> error errMsg
```

Comment:
Perhaps use `clashCompileError` so HDL generation stops properly?

Reply:
Ah, I only read the comment you made last week just now. You can disregard this question.

Reply:
Quote from above:
> [edit]
I temporarily changed the clashCompileError to error so we can look at, for example, the generated Haddock. This term-level treatment will be replaced by a type-level treatment introduced in https://github.com/clash-lang/clash-compiler/pull/2539 anyway.
[/edit]

;-)